coderd/database/dbauthz/dbauthz.go

```diff
@@ -0,0 +1,323 @@
+package dbauthz
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/rbac"
+)
+
+var _ database.Store = (*AuthzQuerier)(nil)
+
+var (
+	// NoActorError wraps ErrNoRows for the api to return a 404. This is the correct
+	// response when the user is not authorized.
+	NoActorError = xerrors.Errorf("no authorization actor in context: %w", sql.ErrNoRows)
+)
+
+// NotAuthorizedError is a sentinel error that unwraps to sql.ErrNoRows.
+// This allows the internal error to be read by the caller if needed. Otherwise
+// it will be handled as a 404.
+type NotAuthorizedError struct {
+	Err error
+}
+
+func (e NotAuthorizedError) Error() string {
+	return fmt.Sprintf("unauthorized: %s", e.Err.Error())
+}
+
+// Unwrap will always unwrap to a sql.ErrNoRows so the API returns a 404.
+// So 'errors.Is(err, sql.ErrNoRows)' will always be true.
+func (NotAuthorizedError) Unwrap() error {
+	return sql.ErrNoRows
+}
+
+func LogNotAuthorizedError(ctx context.Context, logger slog.Logger, err error) error {
+	// Only log the errors if it is an UnauthorizedError error.
+	internalError := new(rbac.UnauthorizedError)
+	if err != nil && xerrors.As(err, internalError) {
+		logger.Debug(ctx, "unauthorized",
+			slog.F("internal", internalError.Internal()),
+			slog.F("input", internalError.Input()),
+			slog.Error(err),
+		)
+	}
+	return NotAuthorizedError{
+		Err: err,
+	}
+}
+
+// AuthzQuerier is a wrapper around the database store that performs authorization
+// checks before returning data. All AuthzQuerier methods expect an authorization
+// subject present in the context. If no subject is present, most methods will
+// fail.
+//
+// Use WithAuthorizeContext to set the authorization subject in the context for
+// the common user case.
+type AuthzQuerier struct {
+	db   database.Store
+	auth rbac.Authorizer
+	log  slog.Logger
+}
```

Comment:
Should we name this something else since the package name is dbauthz? Like `Store`? Or just `Querier`?

Reply:
Does this need to be exposed? Does the interface not suffice?

Reply:
We need the tests in a separate package to avoid an import cycle.
Also, our tests check all the methods on `AuthzQuerier` (which will not always be 1:1 with `database.Store`).

Reply:
@johnstcn we can still get the methods on an unexported struct. We just need to deref the pointer in reflect. (99% sure)

Reply:
Meh, it's fine to expose this IMO, was just curious.

I think the name should be `Querier` though, since the package is called `dbauthz`.